### DISCUSSION TITLE
Are some promise functions guaranteed to complete synchronously?

### DISCUSSION THREAD

**Discussion by PJB3005** [gfx-rs/wgpu-native#260] [Q&A] [answered]
The specification for functions like `requestAdapter()` says they return a promise. On the Rust side this is returned as a `Future`, and on the C header side this is exposed as a callback. All well and good.

However, the [example in the repo](https://github.com/gfx-rs/wgpu-native/blob/bb5d6e9675914c11105621280de3ca8fabaed4ea/examples/triangle/main.c#L233-L238) assumes the call will complete synchronously and have written the result by the time the function returns. I have also seen other tutorials assert that "in practice it will be synchronous".

Is this a guarantee for the future? Or is it simply a shortcut taken by the C example to keep it simple for now?

(Also, is this where these questions go? I couldn't even find where webgpu-headers comes from spec wise, the W3C spec makes no mention of the native header format)

**Comment by almarklein** [comment, marked as answer]
Good question! As I have understood it, the callback will eventually run asynchronous, but the event-loop mechanism has not been fleshed out yet. 

In wgpu-py we also assume it runs synchronous for the time being:
https://github.com/pygfx/wgpu-py/blob/402961a04299a1518ac0fb7ad3dbb44636679dbb/wgpu/backends/rs.py#L280-L282
<img width="730" alt="image" src="https://user-images.githubusercontent.com/3015475/236646678-426012f0-88ac-46a4-896a-350ba9d8c4cf.png">

**Reply by PJB3005** [reply]
If the event loop will have to be pumped from the app side, then I guess I don't want to try doing anything in advance for it. I'll just keep assuming they're synchronous for now too then.
